diag_manager/fms_doubly_linked_list.F90

```diff
@@ -0,0 +1,275 @@
+    class(node), pointer :: nd
+    if(this%theSize /= 0) then
+      nd => this%head%next !!  TODO: verify
+      liter =  this%remove(nd)
+    else
+      liter = this%get_literator()
+    endif
+  end function pop_at_front
+
+  function pop_at_back (this ) result( liter )
+    class(linked_list), intent(in out) :: this
+    class(literator) , allocatable :: liter
+    !!
+    class(node), pointer :: nd
+    if(this%theSize /= 0) then
+      nd => this%tail%prev !! TODO: verify
+      liter = this%remove( nd )
+    else
+      liter = this%get_literator()
+    endif
+  end function pop_at_back
+
+
+
+  !> push (insert) data at the head of the list
+  function push_at_front( this, d ) result(litr)
+    class(linked_list), intent(in out) :: this
+    class(*), target, intent(in out) :: d
+    class(literator), allocatable :: litr
+    litr = this%insert (this%head%next, d)
+  end function push_at_front
+
+  !> push (insert) data at the end of the list
+  function push_at_back( this, d ) result(litr)
+    class(linked_list), intent(in out) :: this
+    class(*), target, intent(in out) :: d
+    class(literator), allocatable :: litr
+    litr =  this%insert (this%tail, d)
+  end function push_at_back
+
+  function node_constructor () result (nd)
+    type(node), allocatable :: nd
+    allocate(nd)
+    nd%data => null()
+    nd%prev => null()
+    nd%next => null()
+    print * , "In node constructor"
+  end function node_constructor
+
+  function linked_list_constructor () result (ll)
+    type(linked_list), allocatable :: ll
+    allocate(ll)
+    allocate(ll%head)
+    allocate(ll%tail)
+    print *, 'associated(ll%head) :' , associated(ll%head), &
+      ' associated(ll%head) :' , associated(ll%head)
+    ll%head%next => ll%tail
+    ll%tail%prev => ll%head
+    print * , "In list constructor"
+  end function linked_list_constructor
+
+
+  function literator_constructor ( fnd, tnd ) result (litr)
+    type (node), pointer  :: fnd
+    type (node), pointer  :: tnd
+    type (literator), allocatable :: litr
+    !type (literator), pointer :: litr
+    allocate(litr)
+    litr%current => fnd
+    litr%end => tnd
+  end function literator_constructor
+
+  function get_size (this) result (sz)
+    class(linked_list), intent(in out) :: this
+    integer  :: sz
+    sz = this%theSize
+  end function get_size
+
+
+  function is_size_zero (this) result (r)
+    class(linked_list), intent(in out) :: this
+    logical :: r
+    if (this%theSize == 0) then
+      r = .true.
+    else
+      r = .false.
+    end if
+  end function is_size_zero
+
+  function get_forward_literator(this) result (litr)
+    class(linked_list), intent(in) :: this
+    class(literator), ALLOCATABLE :: litr
+    litr = literator( this%head%next, this%tail )
+  end function get_forward_literator
+
+
+  !! Should return true unless the currrent is pointing to the next
+  function literator_has_data( this ) result( r )
+    class(literator), intent(in) :: this
+    logical :: r
```

Comment:
The function needs a doxygen description.
Variables need to be documented in doxygen style

Reply:
done!